packages/language/src/language-server/connection-handler.ts

```diff
@@ -45,46 +46,62 @@ export function startLanguageServer(connection: Connection): void {
     };
   });
   connection.onDefinition((params) => {
-    const uri = params.textDocument.uri;
     const position = params.position;
-    const textDocument = TextDocuments.get(uri);
-    const sourceFile = sourceFileHandler.getSourceFile(URI.parse(uri));
-    if (textDocument && sourceFile) {
+    const textDocument = TextDocuments.get(params.textDocument.uri);
+    const uri = URI.parse(params.textDocument.uri);
+    const compilationUnit = compilationUnitHandler.getCompilationUnit(uri);
+    if (textDocument && compilationUnit) {
       const offset = textDocument.offsetAt(position);
-      const definition = definitionRequest(sourceFile, offset);
-      return definition.map((def) => {
-        return {
-          uri: def.uri,
-          range: rangeToLSP(textDocument, def.range),
-        };
-      });
+      const definition = definitionRequest(compilationUnit, uri, offset);
+      const lspDefinitions: Location[] = [];
+      for (const def of definition) {
+        const doc = TextDocuments.get(def.uri);
+        if (doc) {
+          const range = rangeToLSP(doc, def.range);
+          lspDefinitions.push({
+            uri: def.uri,
+            range,
+          });
+        }
+      }
+      return lspDefinitions;
     }
     return [];
   });
   connection.onReferences((params) => {
     const uri = params.textDocument.uri;
     const position = params.position;
     const textDocument = TextDocuments.get(uri);
-    const sourceFile = sourceFileHandler.getSourceFile(URI.parse(uri));
-    if (textDocument && sourceFile) {
+    const parsedUri = URI.parse(uri);
+    const compilationUnit =
+      compilationUnitHandler.getCompilationUnit(parsedUri);
+    if (textDocument && compilationUnit) {
       const offset = textDocument.offsetAt(position);
-      const definition = referencesRequest(sourceFile, offset);
-      return definition.map((def) => {
-        return {
-          uri: def.uri,
-          range: rangeToLSP(textDocument, def.range),
-        };
-      });
+      const definition = referencesRequest(compilationUnit, parsedUri, offset);
+      const lspDefinitions: Location[] = [];
+      for (const def of definition) {
+        const doc = TextDocuments.get(def.uri);
+        if (doc) {
+          const range = rangeToLSP(doc, def.range);
+          lspDefinitions.push({
+            uri: def.uri,
+            range,
+          });
+        }
+      }
+      return lspDefinitions;
```

Comment:
Just a note this duplicates the code above on 55 - 67. We should consider factoring this pattern out later.